### PR TITLE
8299844: RISC-V: Implement _onSpinWait intrinsic

### PR DESCRIPTION
**Pull request by yadongw**
Implement _onSpinWait() intrinsic for RISC-V, and it can be implemented by using PAUSE instruction from Zihintpause Extension, which is ratified and a mandatory extension of RVA22U64 (https://github.com/riscv/riscv-profiles/blob/main/profiles.adoc#rva22u64-mandatory-extensions).
Software can use the PAUSE instruction to reduce energy consumption while executing spinwait
code sequences (https://github.com/riscv/riscv-isa-manual/blob/master/src/zihintpause.tex).

Add a test case of test/hotspot/jtreg/compiler/onSpinWait/TestOnSpinWaitRISCV64.java, passed on QEMU.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299844](https://bugs.openjdk.org/browse/JDK-8299844): RISC-V: Implement _onSpinWait intrinsic

### Reviewers
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11921/head:pull/11921` \
`$ git checkout pull/11921`

Update a local copy of the PR: \
`$ git checkout pull/11921` \
`$ git pull https://git.openjdk.org/jdk pull/11921/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11921`

View PR using the GUI difftool: \
`$ git pr show -t 11921`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11921.diff">https://git.openjdk.org/jdk/pull/11921.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -682,16 +682,20 @@ class Assembler : public AbstractAssembler {
...
-    patch((address)&insn, 27, 24, predecessor);
-    patch((address)&insn, 31, 28, 0b0000);
+    patch((address)&insn, 19, 15, 0b00000);     // rs1
+    patch((address)&insn, 23, 20, successor);   // succ
+    patch((address)&insn, 27, 24, predecessor); // pred
+    patch((address)&insn, 31, 28, 0b0000);      // fm
     emit(insn);
   }
 
+  void pause() {
```
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:695]: https://github.com/openjdk/jdk/pull/10697 moved all pseudo instructions to macroAssembler, I think `pause` should be defined in there too.

**Review by feilongjiang — CHANGES_REQUESTED** [review]
Looks good, with some comments:

File: src/hotspot/cpu/riscv/riscv.ad [hunk 2]
```diff
@@ -1815,6 +1816,8 @@ const bool Matcher::match_rule_supported(int opcode) {
   }
 
   switch (opcode) {
+    case Op_OnSpinWait:
+      return VM_Version::supports_on_spin_wait();
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/riscv.ad:1820]: You can use `UseZihintpause` directly just like for `UseRVV`, `UsePopCountInstruction`, or `UseZbb` below.

**Review by RealFYang — CHANGES_REQUESTED** [review]
Don't forget to update the copyright year.

File: src/hotspot/cpu/riscv/riscv.ad (hunk 2 above)
Comment (yadongw) [reply src/hotspot/cpu/riscv/riscv.ad:1820]: supports_on_spin_wait() is called both from share and cpu specific code. I think it's better to be a unified switch.

File: src/hotspot/cpu/riscv/assembler_riscv.cpp [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /*
  * Copyright (c) 1997, 2012, Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 2014, Red Hat Inc. All rights reserved.
- * Copyright (c) 2020, 2022, Huawei Technologies Co., Ltd. All rights reserved.
```
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/assembler_riscv.cpp:4]: Unnecessary copyright year update of `asembler_riscv.cpp`.

**Review by luhenry — APPROVED** [review]

**Review by feilongjiang — APPROVED** [review]

**Review by RealFYang — APPROVED** [review]
Looks fine to me.

**Pull request closed without merging**